### ISSUE TITLE
MNRL with DDP, how many in-batch negatives will there be?

### ISSUE THREAD

**Issue by OnAnd0n** [UKPLab/sentence-transformers#3149]
I am planning to train an embedding model using MultipleNegativesRankingLoss.

If I set per_device_train_batch_size=4 in the SentenceTransformerTrainingArguments and use 4 GPUs with accelerate for training, 

how many in-batch negatives will there be?

1) in-batch negatives = 15 = 4 * 4 - 1
2) in-batch negatives = 3 = 4 - 1

which is True?

**Comment by tomaarsen** [comment]
Hello!

It depends on if you have `negatives` in your dataset as well!
With `per_device_train_batch_size=4` and 1 `negative` per sample in your dataset, then you have:
| anchor_1 | positive_1 | negative_1 |
|----------|------------|------------|
| anchor_2 | positive_2 | negative_2 |
| anchor_3 | positive_3 | negative_3 |
| anchor_4 | positive_4 | negative_4 |

Here, `anchor_1` is trained to be close to `positive_1`, but far away from `positive_2`, `positive_3`, `positive_4` and all 4 negatives. In short, each anchor (e.g. a query) has 1 positive and 7 negatives here.

If you don't have any `negatives` in your dataset, then you get this:
| anchor_1 | positive_1 |
|----------|------------|
| anchor_2 | positive_2 |
| anchor_3 | positive_3 |
| anchor_4 | positive_4 |

And `anchor_1` is trained to be close to `positive_1`, but far away from `positive_2`, `positive_3`, and `positive_4`. So, each anchor is has 1 positive and 3 negatives. 

The formula is as follows:
Each anchor has 1 positive and (batch_size - 1 + num_negative_columns_in_dataset * batch_size) negatives. 
The `batch_size - 1` portion is for the "other positives" that are used as in-batch negatives, and the `num_negative_columns_in_dataset * batch_size` is for all of the negatives, which are all used as in-batch negatives.

As you can see in the Loss Overview, some loss functions allow for more than 1 negative column. In that case you can get even more in-batch negatives:
![image](https://github.com/user-attachments/assets/ff1afd4e-83eb-49d4-9ff3-85fa43bb2c44)

Additionally, the `Cached...` losses, e.g. [`CachedMultipleNegativesRankingLoss`](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#cachedmultiplenegativesrankingloss), allow you to set a `per_device_train_batch_size` with a lower `mini_batch_size` in the loss, and the memory usage will be based on the `mini_batch_size`. So, if you can only process 4 batches per iteration, then you can get more in-batch negatives by setting e.g. `per_device_train_batch_size=128` and `mini_batch_size=4`. If you e.g. have 1 negatives column, then this means that each anchor has 1 positive and 128 - 1 + 1 * 128 = 255 negatives.

- Tom Aarsen

**Comment by tomaarsen** [comment]
Quick update: As of v5.1, you can use the `gather_across_devices=True` argument in (Cached)MultipleNegativesRankingLoss to also gather negatives across the devices. Works very nicely! More details in the Gathering Across Devices section in https://github.com/UKPLab/sentence-transformers/releases/tag/v5.1.0

I'll close this now, as I think the question's been answered.

- Tom Aarsen
